target_snowflake/__init__.py

```diff
@@ -206,36 +219,62 @@ def persist_lines(config, lines, information_schema_cache=None):
                 raise e
 
             row_count[stream] = 0
+            total_row_count[stream] = 0
             csv_files_to_load[stream] = NamedTemporaryFile(mode='w+b')
+
         elif t == 'ACTIVATE_VERSION':
             logger.debug('ACTIVATE_VERSION message')
+
+        elif t == 'STATE':
+            logger.debug('Setting state to {}'.format(o['value']))
+            state = o['value']
+
         else:
             raise Exception("Unknown message type {} in message {}"
                             .format(o['type'], o))
 
+    # if some bucket has records that need to be flushed but haven't reached batch size
+    # then flush all buckets.
+    if len(row_count.values()) > 0:
+        # flush all streams one last time, delete records if needed, reset counts and then emit current state
+        flush_all_streams(records_to_load, row_count, stream_to_sync, config)
+
+    # emit latest state
+    emit_state(state)
 
+
+def flush_all_streams(records_to_load, row_count, stream_to_sync, config) -> None:
+    """
+    Flushes all buckets and resets records count to 0 as well as empties records to load list
+    :param records_to_load: dictionary with records to load per stream
+    :param row_count: dictionary with row count per stream
+    :param stream_to_sync: Snowflake db sync instance per stream
+    :param config: dictionary containing the configuration
+    :return:
+    """
     # Single-host, thread-based parallelism
     with parallel_backend('threading', n_jobs=-1):
         Parallel()(delayed(load_stream_batch)(
             stream=stream,
             records_to_load=records_to_load[stream],
-            row_count=row_count[stream],
+            row_count=row_count,
             db_sync=stream_to_sync[stream],
             delete_rows=config.get('hard_delete')
         ) for (stream) in records_to_load.keys())
 
-    return state
-
 
 def load_stream_batch(stream, records_to_load, row_count, db_sync, delete_rows=False):
-    #Load into snowflake
-    if row_count > 0:
-        flush_records(stream, records_to_load, row_count, db_sync)
+    # Load into snowflake
+    if row_count[stream] > 0:
+        flush_records(stream, records_to_load, row_count[stream], db_sync)
 
     # Delete soft-deleted, flagged rows - where _sdc_deleted at is not null
     if delete_rows:
```

Comment:
can we DELETE only when `row_count > 0` ?

Otherwise it'll run `DELETE FROM .. stream .. WHERE _sdc_deleted_at IS NOT NULL` for every table even including the ones that hasn't been changed.

Reply:
I agree, let's include the delete in the if statement block.